### PR TITLE
Calibration Optimization Problem Formulation (final final)

### PR DESCRIPTION
**Pull request by h3ndrk**
## Introduced Changes

Introduces camera calibration optimization problem formulation.

## ToDo / Known Issues

None

## Ideas for Next Iterations (Not This PR)

None

## How to Test

LATC (Look at the code)

### CODE REVIEW

File: crates/calibration/src/residuals.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+                        Residuals::calculate_from(parameters, measurements, field_dimensions)
+                            .ok()?
+                            .into(),
+                    );
+                    Some(residuals)
+                }
+                None => None,
+            },
+        )
+        .map(DVector::from_vec)
```
Comment (knoellle) [inline crates/calibration/src/residuals.rs:32]: I like iterators too, but in this case the procedural approach is a lot more readable:
```rs
    let mut residuals = Vec::new();
    for measurement in measurements {
        let vec: Vec<f32> = Residuals::calculate_from(parameters, measurement, field_dimensions)
            .ok()?
            .into();
        residuals.extend(vec.iter());
    }

    Some(DVector::from_vec(residuals));
```

File: crates/types/src/line.rs [hunk 2]
```diff
@@ -95,10 +106,14 @@ impl<const DIMENSION: usize> Line<DIMENSION> {
         self.squared_distance_to_point(point).sqrt()
     }
 
-    pub fn is_orthogonal(&self, other: &Line<DIMENSION>, epsilon: f32) -> bool {
+    pub fn angle_diff_to_orthogonal(&self, other: &Line<DIMENSION>) -> f32 {
         let self_direction = (self.1 - self.0).normalize();
         let other_direction = (other.1 - other.0).normalize();
-        (self_direction.dot(&other_direction).acos().abs() - PI / 2.0).abs() < epsilon
+        (self_direction.dot(&other_direction).acos().abs() - PI / 2.0).abs()
```
Comment (knoellle) [inline crates/types/src/line.rs:112]: We use the premade constants pretty much everywhere else.
```suggestion
        (self_direction.dot(&other_direction).acos().abs() - FRAC_PI_2).abs()
```

File: crates/calibration/src/residuals.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        let distance_between_parallel_line_center_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.center());
+        let distance_between_parallel_line_end_points = projected_lines
+            .border_line
+            .distance_to_point(projected_lines.goal_box_line.1);
+
+        Ok(Residuals {
+            border_to_connecting_angle: border_to_connecting_angle - FRAC_PI_2,
+            connecting_to_goal_box_angle: connecting_to_goal_box_angle - FRAC_PI_2,
```
Comment (knoellle) [inline crates/calibration/src/residuals.rs:80]: Are you sure you know what you are doing?
Manually subtracting angles from each other is dangerous, although I couldn't find any obvious issues in this case.

**Review by knoellle — CHANGES_REQUESTED** [review]

File: crates/calibration/src/residuals.rs (hunk 3 above)
Comment (h3ndrk) [reply crates/calibration/src/residuals.rs:80]: Iirc, I'm using the new angle method that always returns a positive angle between [0,180] degrees. Subtracting 90 degrees transforms the interval into the expected one and gives us a sign for the gradient. I think this should work. 😅

File: crates/calibration/src/residuals.rs (hunk 1 above)
Comment (jonathan-hellwig) [reply crates/calibration/src/residuals.rs:32]: I agree with @knoellle  here

File: crates/calibration/src/residuals.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            distance_between_parallel_line_start_points: distance_between_parallel_line_start_points
+                - field_dimensions.goal_box_area_length,
+            distance_between_parallel_line_center_points:
+                distance_between_parallel_line_center_points - field_dimensions.goal_box_area_length,
+            distance_between_parallel_line_end_points: distance_between_parallel_line_end_points
+                - field_dimensions.goal_box_area_length,
```
Comment (jonathan-hellwig) [inline crates/calibration/src/residuals.rs:86]: The residuals here are measured in different units and ranges. You might want to normalize them or add different weights for each one of them.

File: crates/types/src/line.rs [hunk 5]
```diff
@@ -12,6 +12,17 @@ pub struct Line<const DIMENSION: usize>(pub Point<f32, DIMENSION>, pub Point<f32
 pub type Line2 = Line<2>;
 
 impl Line2 {
+    pub fn signed_angle(&self, other: Self) -> f32 {
+        (self.1 - self.0)
+            .normalize()
+            .dot(&(other.1 - other.0).normalize())
+            .acos()
+    }
```
Comment (jonathan-hellwig) [inline crates/types/src/line.rs:20]: You are not using this function, right? I think you should remove it.

File: crates/types/src/line.rs (hunk 2 above)
Comment (jonathan-hellwig) [inline crates/types/src/line.rs:112]: You can also check whether two vectors are orthogonal by calculating their dot product and checking if the result is zero. That way you do not need to calculate the angle using `acos`.

File: crates/calibration/src/residuals.rs (hunk 4 above)
Comment (tuxbotix) [reply crates/calibration/src/residuals.rs:86]: What I recall from some papers about the solver (Levenberg-Marquardt) is that it is relatively immune to this issue, the ranges of each element of residual vector does not have to be  the same. As I remember the reason is the way it calculates the relative reduction of error - not as a norm or sum but it is done element-wise.

However, as you suggested, weights or some other influencing factors can be introduced to Influence which parameter gets priority, then normalizing would probably help to tune that.

Comment (h3ndrk) [reply crates/calibration/src/residuals.rs:86]: Ok, I will have a look once I conduct more in-depth testing with this calibration problem. I would like to keep it as is for now.

File: crates/calibration/src/lines.rs [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    point: Point2<f32>,
+    which: &str,
+) -> Result<Point2<f32>, LinesError> {
+    matrix
+        .pixel_to_ground(point)
+        .map_err(|source| LinesError::NotProjected {
+            source,
+            which: which.to_string(),
+        })
+}
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/calibration/src/lines.rs:57]: Done

File: crates/calibration/src/problem.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+impl CalibrationProblem {
+    pub fn new(
+        initial_corrections: Corrections,
+        measurements: &[Measurement],
+        field_dimensions: FieldDimensions,
+    ) -> Self {
+        Self {
+            parameters: initial_corrections,
+            measurements: measurements.to_vec(),
```
[the remark this answers is not in the document]
Comment (h3ndrk) [reply crates/calibration/src/problem.rs:26]: Done

File: crates/calibration/src/residuals.rs (hunk 1 above)
Comment (h3ndrk) [reply crates/calibration/src/residuals.rs:32]: Used and improved your suggestion.

File: crates/types/src/line.rs (hunk 2 above)
Comment (h3ndrk) [reply crates/types/src/line.rs:112]: Yep, changed.

File: crates/types/src/line.rs (hunk 5 above)
Comment (h3ndrk) [reply crates/types/src/line.rs:20]: Removed

File: crates/types/src/line.rs (hunk 2 above)
Comment (h3ndrk) [reply crates/types/src/line.rs:112]: Now we have code duplication again but it now does not use `acos()` anymore.

Comment (knoellle) [reply crates/types/src/line.rs]: Careful! You just changed the meaning of the epsilon parameter.
If I see this correctly, you don't even use the `is_orthogonal` function in this PR. It is only used in the field border detection, where an `angle_threshold` parameter is passed for this epsilon.

Furthermore, this implementation isn't even correct.
You just check whether the dot product is less than epsilon without checking for the sign.
If the two vectors point in opposite directions, the normalized dot product would be `-1.0` and the `is_orthogonal` check would return true if we assume epsilon to be some small positive value.

I would suggest reverting this change and renaming the epsilon parameter to something more meaningful.

Comment (h3ndrk) [reply crates/types/src/line.rs]: Reverted, now we also don't have the code duplication anymore. Decided against renaming `epsilon`. Should be clear enough. The function also has been renamed to express the absolute return value.

Comment (jonathan-hellwig) [reply crates/types/src/line.rs]: I am fine with this version.

File: crates/types/src/line.rs [hunk 8]
```diff
@@ -12,6 +12,10 @@ pub struct Line<const DIMENSION: usize>(pub Point<f32, DIMENSION>, pub Point<f32
 pub type Line2 = Line<2>;
 
 impl Line2 {
+    pub fn angle(&self, other: Self) -> f32 {
+        (self.1 - self.0).angle(&(other.1 - other.0))
+    }
```
Comment (h3ndrk) [inline crates/types/src/line.rs:17]: This needs to be reworked. Coming soon.

Comment (h3ndrk) [reply crates/types/src/line.rs:27]: Done

**Review by knoellle — APPROVED** [review]
LGTM

**Pull request merged**